### PR TITLE
Restore previously archived groups

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Adding the ability to restore archived groups and view archived groups in the user groups modal.

I refactored some existing code and converted the user groups modal to a functional component, replaced the infinite scroll mechanism, and added 2 new url parameters to the groups endpoint (`include_archived` & `filter_archived`).

Figma: https://www.figma.com/file/YjrLurbsI40ySBUyl3WdzF/Custom-user-groups-v2?node-id=693-233287&t=zTzmDeOHWe8YwZHE-0
#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-53319

#### Release Note

```release-note
Added 2 new url parameters to GET /api/v4/groups, `include_archived` & `filter_archived`. Added the ability to restore archived groups from the user groups modal.
```

### CODE REVIEW

**Comment by BenCookie95** [comment]
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-22767.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

**Comment by wuwinson** [comment]
@BenCookie95 Found a couple of minor issues, but looks great overall!

1. If a user doesn't have permissions to `Restore deleted user groups`, they still see the "Restore group" button in the group details modal. Should be hidden or disabled. Slight preference it to be hidden for simplicity. Thoughts @abhijit-singh? 
Screen Shot 2023-04-03 at 6 16 09 PM
2. The "G" in "Restore group" should be capitalized in the dropdown and in the group details modal so it's consistent with the rest of the buttons. @abhijit-singh Can you confirm? Mockups show as lowercase. 
Screen Shot 2023-04-03 at 6 15 28 PM

3. Also related to the dropdown above - if a group is archived, the dropdown should only display the option to "Restore Group" based on [requirement 10a](https://docs.google.com/document/d/1HMPjaXIM7cdsF-im9yryAvFEL1Z9w5zYVD2RfQVh8rc/edit#bookmark=id.3fhrr2hj2r06), but I think the current implementation might actually be better. The options menu will always have at least one option to display no matter what permissions are set, so we don't have to worry about hiding or disabling the options menu. What do you think @abhijit-singh?

4. No way to go back from an empty state without closing out and re-opening the modal. Thoughts on adding a back button @abhijit-singh? 

https://user-images.githubusercontent.com/93531870/229640589-9f016e80-e9b4-45d3-b2ac-a9eca9c8e290.mov

5. Dropdowns within the modal get cut-off when there is not at least 4 rows of user groups displayed in the modal. It's not specific to this PR, so let me know if you want me to file this separately @BenCookie95: 

https://user-images.githubusercontent.com/93531870/229641332-19c32788-c8db-4e07-a1b9-8901e93b68ac.mov

**Comment by abhijit-singh** [comment]
Awesome work @BenCookie95! And great points @wuwinson! Adding some feedback:

> 1. If a user doesn't have permissions to Restore deleted user groups, they still see the "Restore group" button in the group details modal. Should be hidden or disabled. Slight preference it to be hidden for simplicity.

Agreed, let's hide the action if the user does not have permissions. This is consistent with the behaviour elsewhere in the product.

> 2. The "G" in "Restore group" should be capitalized in the dropdown and in the group details modal so it's consistent with the rest of the buttons. @abhijit-singh Can you confirm? Mockups show as lowercase.

Let's capitalize it for consistency here, but if I'm not wrong there was a plan to make all in-product copy follow sentence case instead of title case. @justinegeffen can confirm if that's still the plan and if we want to do that here. 

> 3. Also related to the dropdown above - if a group is archived, the dropdown should only display the option to "Restore Group" based on [requirement 10a](https://docs.google.com/document/d/1HMPjaXIM7cdsF-im9yryAvFEL1Z9w5zYVD2RfQVh8rc/edit#bookmark=id.3fhrr2hj2r06), but I think the current implementation might actually be better. The options menu will always have at least one option to display no matter what permissions are set, so we don't have to worry about hiding or disabling the options menu.

Agreed. I've updated the [mockups](https://www.figma.com/file/YjrLurbsI40ySBUyl3WdzF/Custom-user-groups-v2?node-id=631-148708&t=2jEu94gSItVyyQc6-4).

  3a. @BenCookie95 Can we also add an icon to the `Restore group` menu item? It's odd to have one item with an icon and one without. We can use `icon-restore` from [compass-icons](https://mattermost.github.io/compass-icons/).

image


> 4. No way to go back from an empty state without closing out and re-opening the modal. Thoughts on adding a back button @abhijit-singh?

My proposal would be to make the search and filter button persistent here. This is also a problem currently with `My groups` as the filter gets hidden in the empty state if you're not a part of any group. We should update this for both `My groups` and `Archived groups`. Here's the [mockup](https://www.figma.com/file/YjrLurbsI40ySBUyl3WdzF/Custom-user-groups-v2?node-id=1013-252518&t=2jEu94gSItVyyQc6-4):
image

**Comment by BenCookie95** [comment]
> Dropdowns within the modal get cut-off when there is not at least 4 rows of user groups displayed in the modal. It's not specific to this PR, so let me know if you want me to file this separately @BenCookie95:

@wuwinson I made a small adjustment to this but the actual fix will be to migrate to the new menu components, I'll file a separate ticket for that

@wuwinson @abhijit-singh I've addressed all the other issues that were highlighted

**Comment by wuwinson** [comment]
Awesome, thanks @BenCookie95! Last super minor thing, but noticed the text wrapping in the empty state message is slightly different than how @abhijit-singh has it in his design. The text box width should probably be a little smaller so the text wraps more evenly across 2 lines.

Screen Shot 2023-04-18 at 4 30 44 PM

**Review by abhijit-singh — APPROVED** [review]

**Comment by BenCookie95** [comment]
> Looks good to me! Thanks @BenCookie95! Just one tiny nit: The `Restore Group` button on page for an archived group should also probably have the 'G' in `Group` capitalized. image

Fixed

**Comment by BenCookie95** [comment]
> Awesome, thanks @BenCookie95! Last super minor thing, but noticed the text wrapping in the empty state message is slightly different than how @abhijit-singh has it in his design. The text box width should probably be a little smaller so the text wraps more evenly across 2 lines.
> 
> Screen Shot 2023-04-18 at 4 30 44 PM

@wuwinson I've reduced the width to 350px, looks more compact now

**Review by wuwinson — APPROVED** [review]

**Comment by wuwinson** [comment]
Latest changes look good! 👍 

Screen Shot 2023-04-20 at 5 44 22 PM
Screen Shot 2023-04-20 at 5 44 40 PM

**Comment by BenCookie95** [comment]
@calebroseland & @mgdelacroix This is ready for dev review. There are quite a few changes here, let me know if you want me to run you through anything. The biggest changes were:

- Migrating the user groups modal to a functional component and using react window to handle infinite scroll
- Changes to some selectors to make the factory selectors in order to improve some areas where selectors were needlessly re-rendering in the channel post text box
- Added some websockets for when groups are archived/restored

**Review by mgdelacroix — APPROVED** [review]
LGTM, great work!! :rocket: 

A couple of nits noted, nonblocking

**Review by calebroseland — COMMENTED** [review]
Overall looks pretty good. A bit of repetition in some places, and I'm a bit confused why All Groups still shows archived groups, but no other overall concerns.

Should archived groups be showing here?
CleanShot 2023-05-01 at 08 11 56@2x

File: server/channels/store/sqlstore/group_store.go
Comment (BenCookie95) [reply server/channels/store/sqlstore/group_store.go:1551]: When we are including archived groups we need a different type of ordering. Non archived groups should appear first and they are sorted alphabetically, then we display archived groups which are then ordered alphabetically

File: webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx
Comment (BenCookie95) [reply webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx:56]: I'll see if I can quickly refactor. If it ends up being too much I can create a ticket for afterwards

**Comment by BenCookie95** [comment]
> Should archived groups be showing here? CleanShot 2023-05-01 at 08 11 56@2x

Yeah, All Groups will show both non archived and archived groups. Non archived groups should appear first and they are sorted alphabetically, then we display archived groups which are then ordered alphabetically

File: webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx
Comment (BenCookie95) [reply webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx]: I moved the paging stuff into a hook so that state is a bit more readable in the component but I did have to keep the separate states inside the hook

File: webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx
Comment (BenCookie95) [reply webapp/channels/src/components/user_groups_modal/user_groups_modal.tsx]: Good call,  there is only 1 `get*Groups` function now

**Comment by esethna** [comment]
@lindy65 can you help us do a QA on this PR so we can unblock merging?

**Comment by lindy65** [comment]
@esethna - is there a Jira ticket for this PR?  I couldn't fine one...

**Comment by lindy65** [comment]
Hi @esethna 

1. Have E2E tests been written for this feature?
2. I think this needs fixing before merging. It's the first thing customers will notice when they create their first user group:
    - Pop-up that appears when clicking the 3-dots to the right of the user group (when only 1 user group exists) isn't visible
image

**Comment by esethna** [comment]
@lindy65 thanks, I opened a tix here: https://mattermost.atlassian.net/browse/MM-53319

**Comment by sbishel** [comment]
@lindy65 - I updated how the menu dropdown works, please retest.

@calebroseland, @mgdelacroix - Should just need to review my changes for the menu fix.
@BenCookie95 - Nice work...finally ready to be merged.

**Review by lindy65 — APPROVED** [review]
Thanks @sbishel - re-tested as per description in Jira ticket and didn't find any issues.  Menu drop-down is great!

**Review by sbishel — APPROVED** [review]

**Comment by larkox** [comment]
@BenCookie95 I see several changes made on the webapp to (apparently) handle differently certain websocket events, and some changes in Client4. I wonder if any of these changes has to be applied also for mobile. If so, please create a ticket to make sure it is not forgotten and add me as a reviewer when it gets implemented (and probably should merge at the same time as this PR, to make sure there is not any kind of downtime from the mobile app).

File: webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx
Comment (calebroseland) [inline webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx:109]: To prevent upwards cut off

CleanShot 2023-08-22 at 13 32 01@2x


```suggestion
        if (groupListItemIndex === 0) {
```

File: webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx
Comment (calebroseland) [inline webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx:233]: To prevent downwards cutoff

CleanShot 2023-08-22 at 15 21 22@2x
CleanShot 2023-08-22 at 15 36 03@2x


```suggestion
                        height={groups.length >= 8 ? 416 : Math.max(groups.length, 3) * 52}
```

**Review by calebroseland — CHANGES_REQUESTED** [review]
LGTM, just two edge cases when opening menu:
- index=0, length=1
- index=1, length=2

File: webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx
Comment (sbishel) [reply webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx]: fixed https://github.com/mattermost/mattermost/pull/22597/commits/35aef46f275f1bf8f5612629097e346b2c3d209f

File: webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx
Comment (sbishel) [reply webapp/channels/src/components/user_groups_modal/user_groups_list/user_groups_list.tsx]: fixed https://github.com/mattermost/mattermost/pull/22597/commits/35aef46f275f1bf8f5612629097e346b2c3d209f

**Comment by sbishel** [comment]
> @BenCookie95 I see several changes made on the webapp to (apparently) handle differently certain websocket events, and some changes in Client4. I wonder if any of these changes has to be applied also for mobile. If so, please create a ticket to make sure it is not forgotten and add me as a reviewer when it gets implemented (and probably should merge at the same time as this PR, to make sure there is not any kind of downtime from the mobile app).

@larkox - The ABC team is taking over this PR, Ben is off the hook...and on PTO. I think mobile is ok. The changes in Client4 were specifically made backward compatible at the database layer. (https://github.com/mattermost/mattermost/blob/35aef46f275f1bf8f5612629097e346b2c3d209f/server/channels/store/sqlstore/group_store.go#L1586C1-L1590C1).

No additional websocket events were added, but the existing `received_groups` is called in a couple additional places. The mobile app appears to already handle this event [here](https://github.com/mattermost/mattermost-mobile/blob/e3b6ca78d126b69dd22a69d63ae66dcd48a14e18/app/actions/websocket/index.ts#L396)

That said, I added a ticket to ensure mobile testing is completed. https://mattermost.atlassian.net/browse/MM-54217

**Review by calebroseland — APPROVED** [review]

**Comment by sbishel** [comment]
@BenCookie95 - Could you take a look at the check-types error, I'm not sure why its failing?

File: webapp/channels/src/components/team_controller/actions/index.ts
Comment (BenCookie95) [inline webapp/channels/src/components/team_controller/actions/index.ts:59]: @sbishel heads up, I changed `include_archived` to `false` here instead of `true`. On second thought there was no need to for the team controller to fetch archived groups

**Pull request merged**
